test/with_dummyserver/test_https.py

```diff
@@ -301,23 +301,23 @@ def test_assert_specific_hostname(self):
         https_pool.assert_hostname = 'localhost'
         https_pool.request('GET', '/')
 
-    @pytest.mark.xfail
     def test_server_hostname(self):
         https_pool = HTTPSConnectionPool('127.0.0.1', self.port,
                                          cert_reqs='CERT_REQUIRED',
                                          ca_certs=DEFAULT_CA,
                                          server_hostname='localhost')
         self.addCleanup(https_pool.close)
-
         conn = https_pool._new_conn()
-        conn.request('GET', '/')
+        https_pool._start_conn(conn, connect_timeout=None)
 
         # Assert the wrapping socket is using the passed-through SNI name.
         # pyopenssl doesn't let you pull the server_hostname back off the
         # socket, so only add this assertion if the attribute is there (i.e.
         # the python ssl module).
-        if hasattr(conn.sock, 'server_hostname'):
-            self.assertEqual(conn.sock.server_hostname, "localhost")
+        # XXX This is highly-specific to SyncBackend
+        sock = conn._sock._sock
+        if hasattr(sock, 'server_hostname'):
+            self.assertEqual(sock.server_hostname, 'localhost')
```

Comment:
Of course this hack will bite us later... I guess the ideal way to write this test would be to teach the dummy-server to report what SNI name it received? And maybe look at the server cert we got back (which should be accessible through public API)? But I'm happy to worry about that later if you are.

Reply:
Since I forgot to actually say it out loud: what I mean is, I'm leaving this in case you have a thought here, but if you want to go ahead with this as it is then feel free to hit merge.

Reply:
I understood what you mean!

I'd like to fix the issue in this PR, but it could take weeks before I can look at this again (mid january). :(

Reply:
But thanks for being explicit. 👍

Reply:
@njsmith I took a look at this again today!

So it's easy to get the *client* certificate from Tornado by calling [request.get_ssl_certificate()](https://www.tornadoweb.org/en/stable/httputil.html#tornado.httputil.HTTPServerRequest.get_ssl_certificate) and [returning that from GET /certiticate](https://github.com/python-trio/urllib3/blob/bleach-spike/dummyserver/handlers.py#L109-L116). But I don't know how to get the server hostname from Tornado.

I don't know how to get the server cert we got back either, but I don't think it would be really meaningful because the Tornado server we launch only supports one hostname (localhost) anyway.

What do you think?

Reply:
Giving up for now, I added a link to this discussion